README.md
<!-- PROJECT LOGO -->
<br />
<p align="center">
  <a href="https://github.com/MarcusCemes/responsive-image-builder">
    <img src="images/logo.png" alt="Logo" width="80" height="80">
  </a>

  <h3 align="center">Responsive Image Builder</h3>

  <p align="center">
    A modern image processing pipeline
    <br />
    <a href="https://marcus-cemes.gitbook.io/responsive-image-builder/"><strong>Explore the docs »</strong></a>
    <br />
    <br />
    <a href="https://github.com/MarcusCemes/responsive-image-builder/issues">Report Bug</a>
    ·
    <a href="https://github.com/MarcusCemes/responsive-image-builder/issues">Request Feature</a>
  </p>

  [![npm][badge-npm]][link-npm]&nbsp;&nbsp;
  [![Release][badge-release]][link-release]&nbsp;&nbsp;
  [![Node.js][badge-node]][link-node]&nbsp;&nbsp;
  [![Typescript][badge-typescript]][link-typescript]&nbsp;&nbsp;
  ![Make the web lighter][badge-lighter]

</p>


<!-- ABOUT THE PROJECT -->
## About The Project

<div align="center">
  <img width="600" alt="Product screenshot" src="https://gistcdn.githack.com/MarcusCemes/2d76cf16540d85119d70429cd1ff50e0/raw/665941935f000983c77bbb3142712c7e2bb38fc1/Responsive%20Image%20Builder%20-%20Demo%20%2301.svg" alt="An example of usage">
</div>

Responsive Image Builder is a CLI tool to help you automate your image build process. It does this with speed and quality in mind, to give your end-user the best possible experience.

### Built With

* [Node.js](https://nodejs.org/en/) (Typescript)
* [libvips](https://github.com/libvips/libvips) (sharp)

## Motivation

A lighter web is now more important than ever, with cloud computing taking off and their ridiculous bandwidth costs! Your website's images can be the largest source of your outgoing bandwidth, incurring significant costs if not done correctly. Users will thank you for not draining their phone's battery and internet allowance.


## Getting started

<div align="center">

**There is some fantastic documentation for this library, go and check it out!**

[![Read the documentation][badge-docs]][link-docs]

</div>


<!-- CONTRIBUTING -->
## Contributing

Contributions are what make the open source community such an amazing place to be learn, inspire, and create. Any contributions you make are **greatly appreciated**.

1. Fork the Project
2. Create your Feature Branch (`git checkout -b feat/AmazingFeature`)
3. Commit your Changes (`git commit -m 'Add some AmazingFeature'`)
4. Push to the Branch (`git push origin feat/AmazingFeature`)
5. Open a Pull Request

<div align="center">

[![Code of conduct][badge-coc]][link-coc]

</div>

<!-- LICENSE -->
## License

Distributed under the MIT License. See `LICENSE` for more information.



<!-- CONTACT -->
## Contact

Marcus Cemes - [@MarcusCemes](https://twitter.com/MarcusCemes)

Project Link: [https://github.com/MarcusCemes/responsive_image_builder](https://github.com/MarcusCemes/responsive_image_builder)



<!-- ACKNOWLEDGEMENTS -->
## Acknowledgements
* [SHARP](https://github.com/lovell/sharp) - A fantastic wrapper for libvips



<!-- BADGES -->
[badge-npm]:https://img.shields.io/badge/npm-CB3837.svg?style=for-the-badge&logo=npm
[badge-release]:https://img.shields.io/github/release/MarcusCemes/responsive-image-builder.svg?style=for-the-badge&color=FF851B
[badge-travis]:https://img.shields.io/badge/Travis_CI--FFDC00.svg?style=for-the-badge&logo=travis
[badge-node]:https://img.shields.io/badge/Node.js--339933.svg?style=for-the-badge&logo=node.js
[badge-typescript]:https://img.shields.io/badge/Typescript--0074D9.svg?style=for-the-badge&logo=typescript
[badge-lighter]:https://img.shields.io/badge/Make_the_web-lighter-7FDBFF.svg?style=for-the-badge

[badge-docs]:https://img.shields.io/badge/Documentation-007ACC.svg?style=for-the-badge&logo=read-the-docs
[badge-coc]:https://img.shields.io/badge/Code%20of-Conduct-ff69b4.svg?style=for-the-badge

<!-- LINKS -->
[link-npm]:https://www.npmjs.com/package/responsive-image-builder
[link-release]:https://github.com/MarcusCemes/responsive-image-builder/releases/latest
[link-node]:https://nodejs.org
[link-typescript]:https://www.typescriptlang.org
[link-docs]:https://marcus-cemes.gitbook.io/responsive-image-builder/
[link-coc]:https://github.com/MarcusCemes/responsive-image-builder/blob/master/.github/CODE_OF_CONDUCT.md


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/), and this project
adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [4.0.0] - 2020-06-06

Responsive Image Builder has been split into several packages.
This makes it more modular and reduces the installation size.
The core functionality of IPP has been upgraded to process an image pipeline, instead
of fixed functionality based around breakpoint generation.

- @ipp/broker - Task distribution
- @ipp/cli - UI
- @ipp/common - Shared interfaces
- @ipp/core - Image processing
- @ipp/compress - Image compression
- @ipp/primitive - Primitive SVG generation
- @ipp/testing - Shared testing utilities

## [3.1.3] - 2020-01-13

### Changed

- Updated dependencies

### Fixed

- Width/height preset properties now (correctly) accept numbers
- When merging configs, default presets are now overwritten

## [3.1.1] - 2019-06-06

### Changed

- Update dependencies (fixes security vulnerabilities)

## [3.1.0] - 2019-04-26

### Added

- SVG Tracing with _potrace_
- Better modularization of the pipeline for easier extensibility with newer features

### Fixed

- Image converting (`convert` option would do nothing)
- Small formatting errors, typos

## [3.0.1] - 2019-04-02

### Fixed

- CLI ('bin' folder was not whitelisted for deployment)

## [3.0.0] - 2019.04-02

The third major update that involved a rewrite of the worker thread. The image export system is now
more modular and customizable, making it easier to add "filters" in the future to modify image
streams.

This update introduces BREAKING CHANGES to the manifest file and program errors.

### Added

- Image fingerprinting! Don't rely on names anymore for resolution, but on a checksum of the
  original image!
- Temporary files when exporting, less chance of conflicts and overwriting, with negligible
  performance cost
- Better and safer cleanup in case of failed exports
- Custom file naming using template literals!
- A better code-based error system
- Visual Studio Code debug scripts for node `--inspect` debugging (instead of console logging)
- Typescript interface exports
- Code of Conduct + better contributing guide
- Short file header

### Changed

- Rewrote the worker thread to use functions instead of classes, separated into more files, making
  the worker more modular and upgradable in the future.
- The format of the manifest file
- Faster immutable-based undefined value removal (`removeUndefined` replaced with
  `withoutUndefined`)
- Major README rewrite to document the new changes
- Updated dependencies

### Removed

- `WebPictureStyles.ts`, its code was already merged with the react example.

## [2.1.2] - 2019-03-29

### Fixed

- Forgot the `bin` field in package.json, CLI is now usable

## [2.1.1] - 2019-03-23

### Fixed

- WebP optimization
- Typos

## [2.1.0] - 2019-03-23

`imagemin-webp` integration was scrapped as the binary seems to be buggy. The cwebp-bin would
corrupt images with alpha channels, probably due to an incompatibility with `libvips`.

Instead, the quality for `libvips` WebP was reduced to match `imagemin-webp` sizes (q=70).

### Added

- Better WebP optimization
- TIFF support
- `convertFormat` option, for codec conversion
- `exportOriginal` option, for only WebP exports
- More Typescript typings
- Better error codes + documentation

### Fixed

- The "pass-through optimizer function" wasn't returning a promise
- Dual-exports of originally .webp images, causing an error

## [2.0.5] - 2019-03-23

### Changed

- Updated all dependencies. Some now require Node.js 8

### Fixed

- Absolute paths on Windows (tiny-glob bug)

## [2.0.4] - 2019-03-20

### Fixed

- README typos
- Formatted the CHANGELOG

## [2.0.3] - 2019-03-20

### Fixed

- Fixed README layout
- npms.io should now see badges for scoring

## [2.0.2] - 2019-03-20

### Added

- Started keeping a changelog
- semantic-release integration with Travis CI for automatic deployment to GitHub and NPM

### Removed

- standard-version (and other version-related package.json scripts)

### Changed

- docs:(readme) will now trigger a patch release

### Fixed

- Badges are now recognized by the npms.io scoring algorithm

## [2.0.1] - 2019-03-16

### Changed

- Improved README
- Fixed links
- Added new demo image

### Fixed

- Absolute path conflict with `tiny-glob`

## [2.0.0] - 2019-03-16

#### Responsive Image Builder has been re-written from the ground up!

It took a while, but here it is! Bear in mind that this is a major update, introducing many BREAKING
CHANGES. The readme has been re-written, give it a quick check to see what's changed and how to use
the new API.

The TODO list has been extended with future goals, if you have another idea for improvement then why
not fork the project and submit a Pull Request?

### Added

- Typescript
- ES Module import support (tree-shaking is now "possible")
- More image optimizers
- Better documentation
- TODO list
- Travis CI script

### Changed

- Better configuration
- Better performance with Node.js streams

## [1.1.3] - 2018-11-11

### Fixed

- Fixed bug where GIFs would fail to copy from nested directories

## [1.1.2] - 2018-11-08

### Changed

- Use `fs-extra` in favour of `fs`

### Fixed

- Fix bug on UNIX systems when copying from nested directories

## [1.1.1] - 2018-11-04

### Added

- Angular example on how to use the library

## [1.1.0] - 2018-11-04

### Added

- Image optimization

## [1.0.3] - 2018-11-03

First release

[unreleased]: https://github.com/marcuscemes/image-processing-pipeline/compare/v4.0.0...HEAD
[4.0.0]: https://github.com/marcuscemes/image-processing-pipeline/compare/v3.1.3...v4.0.0
[3.1.3]: https://github.com/marcuscemes/image-processing-pipeline/compare/v3.1.1...v3.1.3
[3.1.1]: https://github.com/marcuscemes/image-processing-pipeline/compare/v3.1.0...v3.1.1

lib/worker.js
/** IMPORTS **/

const sharp   = require('sharp');
const path    = require('path');
const fs      = require('fs-extra');
const Export  = require(path.join(__dirname, 'export.js'));
const imagemin = require('imagemin');
const pngquant = require('imagemin-pngquant');
const mozjpeg = require('imagemin-mozjpeg');
const cluster = require('cluster');
let debug; try { debug = require('debug')('RIBworker:' + cluster.worker.id) } catch (err) { debug = () => {} };

let config;
let uncompressed_bytes = 0;
let compressed_bytes = 0;
let raw_bytes = 0;
sharp.concurrency(1);  // Only permit one thread

process.on('message', async function (packet) {
  if (packet.code === 'EXIT') {
    // Send worker stats and exit
    process.send({code: 'SUM', data: {uncompressed_bytes: uncompressed_bytes, compressed_bytes: compressed_bytes, raw_bytes: raw_bytes}});
    process.exit(0);

  } else if (packet.code === 'SETUP') {

    config = packet.data;

  } else if (packet.code === 'ACCEL') {
    sharp.concurrency(config.max_threads);  // Speed up the end, as there are fewer cores

  } else if (packet.code === 'TASK') {
    let result;

    if (!config) {
      result = {code: 'NOPT', data: packet.data};
    } else {

      try {

        exp = await process_image(packet.data, config);

        result = { code: 'OK', data: exp };
      } catch (err) {
        if (err instanceof Error) err = new_error(err);
        result = {code: 'ERROR', data: packet.data, error: err};
      }

    }

    process.send(result);
  }
});


/**
 * Decode and convert the specified image according to passed options
 *
 * @param {String} image_path The path of the image to process
 * @param {Object} options An Options object
 * @returns {Object} An object summarising the export of the image
 * @throws An error if the file was unable to be decoded/converted
 */
async function process_image(image_path, options) {

  const input_path = path.resolve(options.input);               // The input directory
  const output_path = path.resolve(options.output);             // The output directory
  const image_parse = path.parse(image_path);                   // The input image parsed path
  const relative_path = path.relative(input_path, image_path);  // Get the relative path to input/output
  const export_path = path.join(output_path, relative_path);    // Export path
  const export_parse = path.parse(export_path);                 // To remove the extention


  // Try to decode the image metadata, or throw an error
  let metadata;
  try {
    metadata = await sharp(image_path).metadata();
  } catch (err) {
    throw {reason: "Failed to read image metadata", file: image_path, error: new_error(err)};
  }

  // Just copy the file if it's an svg or gif
  if (metadata.format === 'svg' || metadata.format === 'gif') {
    fs.ensureDir(export_parse.dir);
    fs.copySync(image_path, export_path);
    return new Export(image_parse.name, image_parse.ext, 'single', universalPath(relative_path));
  }


  // Try to decode the image and return raw image data
  let image_buffer;
  try {
    image_buffer = await sharp(image_path).raw().toBuffer({resolveWithObject: true});
  } catch (err) {
    throw {reason: "Failed to decode the image", file: image_path, error: new_error(err)};
  }
  raw_bytes += image_buffer.info.size;

  // Process each export sequentially
  const number_exports = options.exports.length;
  let exported_sizes = [];
  let previous_export = null;
  for (let i = 0; i !== number_exports; i++) {

    const current_export = options.exports[i];

    // Calculate whether the export should be processed, or if it will be a duplicate
    let should_export = false;
    if (i === 0 || current_export['force'] === true) {
      // If it's the first export or it should be forced
      should_export = true;

    } else if (previous_export && ((metadata.width > previous_export.width || metadata.height > previous_export.height))) {
      // If it won't result in a duplicate image
      should_export = true;
    }


    if (should_export) {

      let resized_image_buffer;

      // Resize the image
      try {
        const raw_info = {
          raw: {
            width: image_buffer.info.width,
            height: image_buffer.info.height,
            channels: image_buffer.info.channels
          }
        };
        resized_image_buffer = await sharp(image_buffer.data, raw_info).resize(current_export.width, current_export.height, { withoutEnlargement: true, fit: 'inside' }).toBuffer({resolveWithObject: true});
      } catch (err) {
        throw {reason: "Failed to resize the image", file: image_path, error: new_error(err)};
      }
      raw_bytes += resized_image_buffer.info.size;


      // Save it
      try {

        const webp_path = path.join(export_parse.dir, export_parse.name + '_' + current_export.name + '.webp');
        const fallback_path = path.join(export_parse.dir, export_parse.name + '_' + current_export.name + export_parse.ext);

        // Conver to webp
        const raw_info = {
          raw: {
            width: resized_image_buffer.info.width,
            height: resized_image_buffer.info.height,
            channels: resized_image_buffer.info.channels
          }
        };

        // Create the directory tree
        fs.ensureDir(export_parse.dir);

        // Encode to format and save the file, saving the written bytes to a variable
        const webp_uncompressed_size = (await sharp(resized_image_buffer.data, raw_info).webp().toFile(webp_path)).size;
        const fallback_uncompressed_size = (await sharp(resized_image_buffer.data, raw_info).toFormat(metadata.format).toFile(fallback_path)).size;

        raw_bytes += webp_uncompressed_size + fallback_uncompressed_size;

        // Apply image optization
        if (config.optimize === true) {
          if (metadata.format === 'png') {

            await imagemin([fallback_path], export_parse.dir, { use: [ pngquant({ quality: "70-90", floyd: 1, speed: 1 }) ] });
          } else if (metadata.format === 'jpeg') {
            await imagemin([fallback_path], export_parse.dir, { use: [ mozjpeg({ quality: 80 }) ] });
          }
        }

        // Store statistical data
        compressed_bytes += fs.statSync(fallback_path).size + webp_uncompressed_size;
        uncompressed_bytes += webp_uncompressed_size + fallback_uncompressed_size;

        // Push the export to the exported_sizes array
        const to_push = {name: current_export.name, width: current_export.width, height: current_export.height};
        to_push['webp'] = universalPath(path.relative(output_path, webp_path));
        to_push['fallback'] = universalPath(path.relative(output_path, fallback_path));
        current_export['default'] === true ? to_push['default'] = true : null;

        exported_sizes.push(to_push);

      } catch (err) {
        throw {reason: "Failed to save the image", file: image_path, error: new_error(err)};
      }

      resized_image_buffer = null;  // Free memory

    }

    previous_export = current_export;

  }

  image_buffer = null;  // Free memory

  // Export complete
  return new Export(image_parse.name, image_parse.ext, 'multiple', null, exported_sizes);

}

function new_error(err) {
  return { name: err.name, stack: err.stack, fileName: err.fileName, lineNumber: err.lineNumber, columnNumber: err.columnNumber};
}


/**
 * Split the path and stitch it up
 * exclusively with '/', regardless of
 * the operating system.
 *
 * @param {*} p A RELATIVE path (NO ROOT)
 * @returns A path seperated with '/'
 */
function universalPath(p) {
  return p.split(path.sep).join('/');
}
